src/storage/src/source/mod.rs

```diff
@@ -543,16 +538,34 @@ impl fmt::Debug for SourceMessage<Option<Vec<u8>>, Option<Vec<u8>>, ()> {
     }
 }
 
+/// A batch of messages from a source reader, along with the current upper.
+pub struct SourceMessageBatch<Key, Value, Diff> {
+    messages: HashMap<PartitionId, Vec<(SourceMessage<Key, Value, Diff>, MzOffset)>>,
+    /// Any errors that occured while obtaining this batch.
+    // TODO: These non-definite errors should not show up in the dataflows/the
+    // persist shard but it's the current "correct" behaviour. We need to fix
+    // this as a follow-up issue because it's a bigger thing that breaks with
+    // the current behaviour.
+    non_definite_errors: Vec<SourceError>,
+    source_upper: HashMap<PartitionId, MzOffset>,
```

Comment:
probably `new_source_upper` is a better name, as we are communicating a new, non-remapped upper

Reply:
I'm risking bikeshedding but I think we shouldn't change the name based on a process (remapping) that happens later and is unrelated to a source producing batches